### DISCUSSION TITLE
Pull request page become so slow when pr is big

### DISCUSSION THREAD

**Discussion by namse** [community/community#33663] [Pull Requests]
![image](https://user-images.githubusercontent.com/3580430/191431916-e0ba8620-7965-423f-bcac-08750d0b8d3c.png)

My company used to make a big PR for major update. It's very difficult to review or check comment with code in Pull request page because it become too laggy to review in page.

I hope it become optimized.

**Comment by robotocryptogame** [comment]
Same problem here. Is horrible. this should be high priority

**Reply by aryanbrite** [reply]
❤️

**Comment by GoodPie** [comment]
I've been using the `Github Pull Requests` extension to get around this at the moment. With a relatively large feature pull request (~50 files touched), starting a comment causes the whole page to become unresponsive. With the extension, it seems to be a lot smoother.

**Comment by james-arawhanui** [comment]
Same here on Chrome. Switching to Firefox, and it's no longer an issue.

Here's [my tampermonkey script](https://gist.github.com/james-arawhanui/66bfb3f76491d11384d9bf13ab843272) to remind me to use firefox instead.

**Comment by hamzaseif13** [comment]
Same here

**Comment by royathan** [comment]
This has gotten incredibly bad.

**Comment by nbolton** [comment]
Above 50 files changed the UI becomes unusable. How about paging the results instead of loading more than the browser can handle?

**Comment by timofeymukha** [comment]
Indeed, it is horribly slow

**Reply by nbolton** [reply]
What browser?

**Reply by royathan** [reply]
It's rough in Chrome (massive slowdowns) it's unusable in Safari (can't modify PR description text).

It's reproducible by opening a PR in a go project that vendors the dependencies, you'll get a massive million line pr.

**Comment by ujain6** [comment]
It is so bad, can't believe a company at GitHub's level is doing this

**Reply by nbolton** [reply]
What browser?

**Reply by MarkMurphy** [reply]
> What browser?

Safari (Version 18.3 (20620.2.4.11.5)) on MacOS (Apple M4 Pro) for me. I also have no extensions installed.

**Reply by thomas-bc** [reply]
Using Chrome on both Apple M4 Pro and Apple Intel 2019 on my end - same issue

**Comment by thomas-bc** [comment]
Commenting to get more traction on this. Using Chrome - PRs aren't even that big (~50-80 files, not too many lines per file) and slowness becomes unbearable. It really does sound like paging could help ?

**Reply by jameswilliamknight** [reply]
Here's a few projects that may shed some light, in case it's react rerenders.

https://github.com/aidenybai/bippy

https://github.com/aidenybai/react-scan

I saw some magic using this:

https://www.npmjs.com/package/babel-plugin-react-compiler

**Reply by nbolton** [reply]
It would be useful to know if you also see the issue in Firefox.

**Reply by thomas-bc** [reply]
@nbolton ah that's interesting. The issue is almost gone when using Firefox. Still a little slow to load, but nothing unexpected for a large page like this. Chrome is incredibly painfully slow.

**Reply by nbolton** [reply]
I think the GitHub devs only work in Firefox :p

**Reply by danielschenk** [reply]
@jameswilliamknight Actually, this video shows that enabling the compiler can indeed fix this easily. GitHub is shown as a specific example of the rerender issue. 

https://youtu.be/INLq9RPAYUw?si=ExVnt2mrQBvTMG4e

**Comment by nbolton** [comment]
@namse Please could you rename your discussion to:
"Pull request page become so slow on Chrome when PR is big"

I think the issue is Chrome specific, and the GitHub devs should know that.

**Reply by namse** [reply]
Would you give me reasonable data or benchmark-thing that chrome is the only browser that have a lag on big pr?

**Reply by nbolton** [reply]
What browser are you using? It appears to be Chrome from your screenshot.

Of the approx 5 people I asked in this discussion, and the other larger discussion, #39341, all have said they use Chrome. Small sample size but quite telling so far.

Two of those people and myself observed that using Firefox solves the issue.

This is why I think the issue is Chrome-specific.

**Reply by Myloc-MEK** [reply]
For me Chrome is worse than Firefox but Firefox also starts lagging when PRs get bigger. The time to post a review comment or mark a file as viewed also increases with bigger PRs or when the number of pending review comments are more than about 50.

**Reply by hamzasharuf** [reply]
I'm experiencing the same issue on Safari, so it's not Chrome-specific issue

**Reply by namse** [reply]
I fundamentally suspect that this issue arises because the GitHub PR viewer loads HTML tags for all files at once, overloading the browser. This can be seen as a problem similar to implementing infinite scroll. Not only GitHub but also other websites have experienced significant lag when displaying a large amount of content within a scrollable view.  

There may be differing opinions on whether this optimization issue should be resolved by the browser or by the website (GitHub). However, one thing is clear—if GitHub wants to provide a better experience for its users, they should take the initiative. Therefore, I don't want to attribute this issue to a specific browser unless sufficient data can support that conclusion.

**Reply by nbolton** [reply]
> I'm experiencing the same issue on Safari, so it's not Chrome-specific issue

Good to know. Did you try running Safari without extensions?
https://apple.stackexchange.com/questions/173647/is-it-possible-to-disable-safari-extensions-when-using-private-browsing

> For me Chrome is worse than Firefox but Firefox also starts lagging when PRs get bigger

How about when you run Firefox without add-ons?
https://smallbusiness.chron.com/start-firefox-addons-59185.html

**Reply by Myloc-MEK** [reply]
It seems to be the plugin for my password manager that is causing the poor performance on Firefox, if i disable it and refresh the page it feels like it is working without lag on a PR with 300+ changed files, I will check later if it's the same when I have added multiple comments, the same can not be said for Chrome, I tried running Chrome in incognito mode and also in regular mode with all extensions disabled and it still had bad performance

**Reply by nbolton** [reply]
> if i disable it and refresh the page it feels like it is working without lag on a PR with 300+ changed files

Amazing! Great testing. I didn't try with Firefox add-ons but good to know that add-ons contribute to the problem.

> the same can not be said for Chrome, I tried running Chrome in incognito mode and also in regular mode with all extensions disabled and it still had bad performance

Yep, same here.

**Comment by nbolton** [comment]
# Workaround

**Temporary solution until GitHub fixes this:** Use Firefox with _add-ons disabled_.

Many people have reported that Firefox (with add-ons disabled) has far better performance on the PR file tab compared to Chrome, Chromium and Safari. Many say that disabling extensions in Chrome doesn't seem to help (I can confirm this).

**Reply by eduardoldc** [reply]
Using Edge is also another possible solution, since I couldn't use Firefox I decided to try it and so far it is working much better than Safari/Chrome for me.

**Comment by peppy** [comment]
Here's [an example of a PR](https://github.com/ppy/osu/pull/31774) that isn't even too large causing stalls of 2-10 seconds on Safari (Mac Studio, highest specs) when performing operations such as as scrolling (causing the header content to anchor to the top of the page), changing tabs, typing in a textbox, etc.

For tab changes, there is a large overhead of 5 second spent in JS (turbolinks). For scrolling, style invalidation is 2-4 seconds. Typing the first character in a textbox has a noticeable delay. Posting review feedback can sometimes get in a stuck state where the content is posted but the page never reports that it is, likely due to low performance.

Please github give this some attention.

**Reply by jameswilliamknight** [reply]
Great testing. I tested by opening that PR in Zen Browser (Firefox) also, and no noticeable lag; plugins:

<details>
  <summary>Browser Extensions List</summary>

  - Awesome Screenshot  
  - Bitwarden Password Manager  
  - Dark Reader  
  - Ghostery Tracker & Ad Blocker  
  - Momentum  
  - Obsidian Web Clipper  
  - React Developer Tools  
  - Reddit Enhancement Suite  
  - Redux DevTools  
  - SponsorBlock for YouTube  
  - Tampermonkey  
  - uBlock Origin  

(please feel free to tell me if any of these are garbage and/or malicious)
</details>

**Reply by peppy** [reply]
Switching from "files changes" to "conversation" takes around 11 seconds, with the majority being CPU bound:

![Safari 2025-02-14 at 03 44 03](https://github.com/user-attachments/assets/2403e236-41fd-426e-990f-02944fdbead3)

![Safari 2025-02-14 at 03 45 28](https://github.com/user-attachments/assets/0ba0960b-71d0-4d00-b7c8-2264a4577771)

This is just one example, but currently the most egregious I can find.

**Reply by laurencee** [reply]
Looks like this has been fixed as it loads fine on firefox and chrome for me as of today at least (tested on Win10 & 11). It could of course be faster, but it doesn't freeze the whole browser.

I've tried some larger PRs too with 100 odd files, 5k lines changes, and it also loaded fine with a loading bar/progress for around 5 seconds rather than completely locking the browser.

Might have been a very recent fix.

**Comment by thomas-bc** [comment]
Safari and Chrome are the [two largest browsers by market share](https://gs.statcounter.com/browser-market-share), and by far. It does not sound right that such a large issue would be ignored by GitHub, and the only workaround is "use something else"...

**Reply by pabloarista** [reply]
it has def gotten worse this past year and in the last few days too. I thought something was wrong with my machine when another team member told me they have the same issue and only with GitHub. FF is a nice workaround because I need to be productive, but pretty sad that all of MS's products are just garbage lately. I haven't used windows as my main machine in years, I wonder how it is there too. I just remember they had some solid products when I was a windows dev. what is going on with this company? Too busy working on AI and neglecting everything else?

**Reply by peppy** [reply]
I believe things got to a new breaking point with the recent UI component changes that github rolled out, based on the sudden influx in complaints in this (and other threads), and also based on my personal perception – crossing the threshold of wanting to continue working on a pull request review.

**Comment by HanMoeHtet** [comment]
I look forward to this issue getting solved soon.
Please forgive my bad language in the previous comment as I was stressed and losing my temper while writing it.

**Reply by bhagdave** [reply]
This is also an issue on Firefox.

**Comment by bene-tyler** [comment]
Ditto on all the comments. Large PRs freeze most of our teams browsers.

**Comment by vsmid** [comment]
A truly horrible user experience, I can't believe this has not been fixed already. GitHub please?

**Comment by SystemInvecklare** [comment]
Please Github! We're dying over here

**Comment by Prasundas99** [comment]
20 File changes is also not showing on any Chromium browsers

**Comment by njfamirm** [comment]
still have issue, if pr is very large, 100 files and more

**Comment by Jgordon-pencilwrench** [comment]
Adding fuel to the fire. The UI is approaching unusable. I'm on the latest Safari/MacOS.

**Comment by thomas-bc** [comment]
It makes no sense to me that GitHub would not even acknowledge this and at least pretend to look into it.

**Reply by Prasundas99** [reply]
They are too busy working with AI, doing vibe coding

**Comment by damonearl** [comment]
I am constantly crashing in Safari while doing reviews, GitHub needs to fix this it's clearly getting worse overtime.

**Comment by udlose** [comment]
I'm hoping this release finally resolves this issue! https://github.blog/changelog/2025-06-26-improved-pull-request-files-changed-experience-now-in-public-preview/

**Reply by minoesteban** [reply]
Tried it just now by enable it via Feature Preview (Safari, latest version) - improves the experience by a lot

**Reply by peppy** [reply]
Can also confirm that things are slightly better with the new version. There's still issues but largely alleviated to the point it's no longer frustrating.

**Reply by mariczne** [reply]
No wonder things seem better:

> Warning
> Only the first 300 files are currently being shown. To see more, switch back to the classic page.

**Reply by peppy** [reply]
We were seeing issues with far less than 300 files, and those cases are looking much better.

**Reply by themightyoarfish** [reply]
For me this happens even with <40 files, none of which are particularly big. I don't know if this is a firefox browser issue yet, but it's pretty much unusable for anything but really small PRs.

**Comment by malekelkssas** [comment]
The issue is even worse when using the Edge browser.

**Comment by Celtian** [comment]
This is so bad. Show us that React is better than Angular. IMHO github should be the proof that React is useful for large scale apps, but it is not.

**Comment by znorris** [comment]
Very slow, freezes my Windows machine when using the PR review site. Resizing the window is a sure-fire way to freeze my system now.

**Comment by SAGV** [comment]
Ohhh, we've switched to Github from Gitlab for our corporate repos and… wow, so hard to review PRs. The page is literally dying on us when using Arc browser (a fresh fork of chrome). 

Come on, Github, you are a huge company! How do you even manage to review your own PRs? 🤣

**Comment by themightyoarfish** [comment]
This occurs on Firefox too, its unusable quite often for me.

**Comment by crash83k** [comment]
I highly doubt this would be an issue if Microsoft hadn't bought GitHub. It's been down hill since.

What's even crazier to me is that this issue is going on 3 years now, and being the LARGEST open source software repository source in the world, you'd imagine that there'd be more effort to try and fix this.

3 years next month since this discussion started!!!

It's obvious that, because browsers limit their resource usage, that the specs of your computer will not really matter here because GitHub is already hitting the resource maximums with just loading the page. **face palm**

I'm sure there's a lot of engineers trying to come up with work-arounds and bandaids, but this needs to be fixed at the root of the issue.

**Comment by sanamhub** [comment]
GitHub now really needs a fierce competitor (and something will come soon), because they're not fixing fundamental issues of software engineering and modern software.

https://yoyo-code.com/why-is-github-ui-getting-so-much-slower/

**Reply by crash83k** [reply]
Excellent write up.

**Comment by njfamirm** [comment]
I think a change happen in new pr mode, now Github use lazy render and don't render all files at same time, in pr with 90 files, with +4000 lines, the issue not exists

**Comment by eduardoldc** [comment]
Last couple weeks I did not have problems on Safari but now I would say it is even worse, to the point that I am not even able to approve a 30 files PR

**Comment by adomaitisc** [comment]
Using Helium Browser - Its chromium... and the pr page completely sucks

**Comment by delyan-peev** [comment]
This is absolutely atrocious, I'm barely able to comment on a 12-file PR at this point (chromium browser). Is this really being ignored for that long?

**Comment by liqwid** [comment]
Goes to show that most likely vast majority of GitHub's paying user base never does code reviews 🥲

**Comment by emilte** [comment]
I experience loading spinners in the conversation tab. It's weird that elements that have already been rendered is struggling to display when i don't even interact with anything other than scrolling.

**Comment by CyanoFresh** [comment]
same, chrome macos m2

**Comment by johndelong** [comment]
We are experiencing incredibly slow load times on even medium sized PRs (30 some files). Often it's not even reviewable in Github and we end up looking at the git commits and reviewing the code in our IDEs. Review in Github has really gone down hill lately where its painful to use now. Using Safari on MacOS M2 machine.

**Comment by thejollyrogers** [comment]
This is horrible Github, this is all we use your platform for is code reviews. Can anyone recommend alternatives?

**Reply by jameswilliamknight** [reply]
GitLab. 

https://handbook.gitlab.com

**Comment by ignatremizov** [comment]
Cmon Github, the UI has become horrifically slow. This should be a top priority

**Comment by StitiFatah** [comment]
Literally hit a "504 Gateway Time-out" and a "We couldn't respond to your request in time." right now while trying to write this message, not even talking about that blue progress bar that you see forever. 

But who cares right ? The most important thing is that Github is "an AI first platform" !  You can even use AI to fix vulnerabilities in your code to help your "AI empowered team" !!!

**Comment by thad0ctor** [comment]
2026 and still an issue...

**Comment by elyemit** [comment]
We face the same issue, it is extremely slow on large PRs.

**Comment by plxel** [comment]
C'mon Github! I believe in you

We moved recently from Gitlab to Github, Github is much better in a lot of things, but this is embarrassing issue to have.

How this can live for years?

**Comment by plxel** [comment]
Looks like there is some new view toggle, which seems working better, probably with its own problems, but at least performance is better
<img width="267" height="124" alt="image" src="https://github.com/user-attachments/assets/2ca80fe6-091d-42b1-b1f5-f4dace5055e0" />

**Comment by AtharvaVaidya** [comment]
Same here, please fix

**Comment by cuquo** [comment]
I just came here to congratulate the team. The new experience is dramatically faster than the classic one. Memory usage is much lower, and render speed is significantly faster. The only issue I’ve noticed is some cumulative layout shift while scrolling, but overall it feels very snappy.

**Comment by imanmahjoubi** [comment]
Hi everyone 👋🏻 ,I'm a PM on the PRs team. I wanted to share that we recently shipped major performance updates to the new Files changed experience.

 Pull request diffs are now more responsive when clicking, typing, and scrolling, with the largest pull requests seeing the biggest gains. For the largest pull requests, the virtualized experience is now enabled by default to optimize performance, reduce memory usage, and improve stability. It’s also much faster to navigate between the Conversations and Files tabs, file tree resizing is smoother and more responsive even for large pull requests, and Safari performance and stability have been improved. Please let us know if you’re still experiencing any performance issues or noticing anything that feels slow. 
 
 Thank you!

**Reply by mikedsharp** [reply]
For the past few days i've been seeing the Unicorn when trying to view a particularily large Pull Request.

Before it was slow, now my team can't even see the file changes to review the pull request, or get to the "review changes" button

**Reply by RobertKrawitz** [reply]
I would still like the option to view the entire diff in one view (not just a window into the PR).  I have a powerful system with a lot of memory (and swap) and am willing to take the chance on things running very slow and even crashing my browser (and shortening the life of my swap drive).

**Comment by mikedsharp** [comment]
Anyone else just seeing the "This page is taking too long to load." message when viewing a large Pull Request now?

I was previously able to view the PR, albeit with many of the diffs would refuse to load, but you could still add reviews and comments.

Now since two days back, I can't even view any of the changes because of the "This page is taking too long to load." page always shows.

**Comment by GringoDotDev** [comment]
This is still horribly slow, even on PRs that are moderately big rather than yuge.

**Comment by hratio** [comment]
Disabling Bitwarden form-auto-fill removed all my CPU issues with huge PR's. 

Seems like some extensions have a Observer that does querySelectorAll on the whole page when scrolling and whatnot.

**Reply by Eduardo-Kaluf** [reply]
I'm currently reviewing a huge PR with over 300 files, and this completely solved the problem. Thank you!

**Reply by Meakk** [reply]
Same here, that was the problem, thanks!

**Reply by MalekSmida** [reply]
Same here thanks !
I am on Firefox.

**Reply by ganor38** [reply]
You saved my life

**Reply by darkfusion90** [reply]
No longer facing slowdowns after this. Thank you so much!

**Reply by abhinavxd** [reply]
this worked

**Comment by tedwards-sf** [comment]
This is still a problem for me, using latest Chrome in Windows 11. When viewing a moderately large diff in a PR and scrolling down, my computer becomes unresponsive for at least a few seconds, sometimes much longer. After that it's ok, until I leave it for a while and do it again
